### ISSUE TITLE
接入 MLflow 并统一 PPO 训练监控与指标管理

### ISSUE THREAD

**Issue by xcz0** [xcz0/Eco-AutoDrive#93]
## 背景

当前 PPO 训练已经具备较完整的训练诊断数据，包括 loss、KL、clip fraction、entropy、gradient norm、reward components、Beta policy statistics、速度/进度/安全/能耗等指标，并通过 `TrainingUpdateSummary` + `update_observer` 暴露更新级监控接口。

项目同时已经使用 Lightning Fabric 管理 device、precision、seed、model setup 和 training checkpoint，PPO 算法部分则基于 TorchRL 实现 GAE、ClipPPOLoss 和 replay buffer。

当前问题主要是训练过程仍缺少统一的实验跟踪和可视化入口，部分指标聚合也依赖手动的 `sum/mean/max/count` 实现。计划引入 MLflow，并进一步利用 Lightning / TorchMetrics 减少外围训练基础设施和指标统计的手工代码。

## 总体思路

职责划分：

- **TorchRL**：继续负责 PPO、GAE、policy ratio、clipped loss 等 RL 算法逻辑。
- **TorchMetrics**：逐步负责 reward、speed、energy、collision rate、loss mean、gradient norm 等通用指标聚合。
- **Lightning Fabric**：继续负责 device、mixed precision、seed、module setup、checkpoint，并作为统一 logging 入口。
- **MLflow**：负责 experiment/run 管理、超参数记录、训练曲线、artifact 和不同实验间比较。
- **Eco-AutoDrive**：保留 reward、MetaDrive rollout、energy evaluation、policy guidance 等领域逻辑。

目标数据流：

```text
MetaDrive rollout
    -> TorchMetrics / rollout statistics
    -> TorchRL PPOUpdater
    -> TrainingUpdateSummary
    -> Fabric logging
    -> MLflow
```

### 1. 接入 MLflow

为每次训练建立独立 MLflow Run，并统一记录：

- Params
  - PPO config
  - reward profile / reward weights
  - sampler config
  - runtime seed
  - scenario / protocol 信息
- Metrics
  - `ppo/*`: policy/value/entropy/total loss、KL、clip fraction、entropy、explained variance、policy ratio、learning rate
  - `gradient/*`: pre-clip gradient norm 及已有分组 gradient diagnostics
  - `reward/*`: mean total/base reward、safety gate
  - `reward_component/*`: TTC、progress、comfort、speed、energy
  - `policy/*`: alpha/beta、action statistics、boundary mass
  - `behavior/*`: speed、progress、stopped fraction、collision、out-of-road
  - `energy/*`: fuel proxy / ml_per_km 等
- Artifacts
  - resolved Hydra config
  - `summary.json`
  - runtime metadata
  - selected policy checkpoints
- Tags
  - git commit / branch
  - experiment ID，例如 A0/A1/A2
  - reward profile

保留当前 `summary.json + NPZ + checkpoint` 作为正式 research artifacts，MLflow 作为实验索引、比较和可视化层，不替代现有 artifact contract。

### 2. 利用 Lightning Fabric 统一 logging

当前 Fabric 已负责 runtime 和 checkpoint，逐步把训练侧日志统一到一个 logging 边界，例如：

```text
TrainingUpdateSummary
    -> metrics adapter
    -> Fabric log_dict
    -> MLflow backend
```

避免在 `PPOUpdater`、rollout collector、reward evaluator 等模块中直接散落 `mlflow.log_metric()`。

现有 `update_observer` 可以作为过渡接口，第一阶段直接将其实现为 MLflow observer，后续再视需要收敛到 Fabric logger。

### 3. 引入 TorchMetrics 减少手动聚合

优先替换通用统计逻辑，而不改 PPO 数学实现。

适合迁移到 TorchMetrics 的内容：

- reward mean / sum
- speed mean
- progress mean
- energy mean
- stopped fraction
- collision / out-of-road rate
- PPO loss mean
- KL / entropy mean
- gradient norm mean/max
- reward component mean

可按 `RolloutMetrics`、`PPOMetrics`、`RewardMetrics` 等 `MetricCollection` 组织。

暂不迁移：

- GAE
- ClipPPOLoss
- PPO policy ratio
- Beta distribution
- actor/value/shared trunk gradient diagnostics
- domain-specific reward / energy computation

这些继续保留在 TorchRL 或项目自身实现中。

## 可视化重点

MLflow Dashboard 优先关注以下几组曲线：

1. PPO optimization：loss、KL、clip fraction、entropy、explained variance、learning rate。
2. Gradient：pre-clip gradient norm、分组 gradient diagnostics。
3. Reward：mean total/base reward 和各 reward component。
4. Policy：Beta alpha/beta、action mean/std、boundary mass。
5. Closed-loop behavior：speed、progress、stopped fraction、collision/out-of-road。
6. Energy：ml/km，并与 speed/progress 同时观察，避免低能耗来自低速或停止等 reward hacking。

训练指标和正式 evaluation 指标继续区分。MLflow 可以同时记录 checkpoint evaluation 结果，但 held-out evaluation 不直接参与训练侧 reward 统计。

### Phase 1: MLflow baseline

- 增加 MLflow dependency/config。
- 将现有 `TrainingUpdateSummary` 直接记录到 MLflow。
- 记录 Hydra config、Git metadata、summary 和关键 checkpoints。
- 支持 A1/A2、多 seed、多 reward weight 的 run 对比。

### Phase 2: TorchMetrics cleanup

- 识别 `build_update_summary` / rollout 中重复的 mean/sum/max/rate 聚合。
- 用 TorchMetrics 替换通用统计逻辑。
- 保持现有 `TrainingUpdateSummary` schema 尽量稳定，避免影响实验 artifact。

### Phase 3: Fabric logging consolidation

- 评估通过 Fabric logger / `log_dict` 统一指标发送。
- 减少直接 MLflow API 调用，使 MLflow 成为可替换 backend。
- 保留当前 custom PPO training loop，不引入 Lightning Trainer。

## 验收标准

- [ ] 单次 PPO 训练在 MLflow 中生成独立 Run。
- [ ] 可实时查看 loss、reward、gradient、policy、behavior、energy 主要指标。
- [ ] Hydra/PPO/reward/runtime 配置可在 MLflow 中检索和比较。
- [ ] A1/A2 或不同 seed 的训练曲线可直接进行横向对比。
- [ ] `summary.json`、runtime metadata 和关键 checkpoint 被关联到对应 Run。
- [ ] 接入 MLflow 后不改变 PPO、reward 和 rollout 的数值语义。
- [ ] 通用指标聚合逐步迁移到 TorchMetrics，但 TorchRL PPO 实现保持不变。
- [ ] 保持现有 resume/checkpoint 行为和训练可复现性。
